File: Algorithms/kadane's_algorithm.cpp
#include<bits/stdc++.h>
using namespace std;

// Function to implement Kadane's Algorithm to find the maximum subarray sum
int kadanes_algo(int n, vector<int>& temp) {
    int max_sum = INT_MIN;  // Initialize max_sum to the smallest possible value
    int curr_sum = 0;       // Initialize current sum to 0

    // Loop through the array
    for (int i = 0; i < temp.size(); i++) {
        curr_sum += temp[i];              // Add the current element to curr_sum
        max_sum = max(max_sum, curr_sum); // Update max_sum if curr_sum is greater

        // If curr_sum becomes negative, reset it to 0 (we ignore negative subarrays)
        if (curr_sum < 0) {
            curr_sum = 0;
        }
    }
    return max_sum;  // Return the maximum subarray sum
}

int main() {
    int n;
    cin >> n;  // Read the size of the array

    vector<int> temp;  // Create a vector to store the array elements

    // Read n elements into the temp vector
    for (int i = 0; i < n; i++) {
        int arr;
        cin >> arr;     // Input each element
        temp.push_back(arr);  // Add the element to the vector
    }

    // Output the result of Kadane's Algorithm
    cout << kadanes_algo(n, temp);

    return 0;
}

File: Algorithms/Kadane's Algorithm.java
// Java program to implement Kadane's Algorithm
public class KadanesAlgorithm {

    // Function to find the maximum sum of a contiguous subarray
    public static int maxSubArraySum(int[] arr) {
        int maxSoFar = Integer.MIN_VALUE; // To store the maximum sum so far
        int maxEndingHere = 0;            // To store the sum of the current subarray

        for (int i = 0; i < arr.length; i++) {
            maxEndingHere += arr[i];

            // Update maxSoFar if maxEndingHere is greater
            if (maxSoFar < maxEndingHere) {
                maxSoFar = maxEndingHere;
            }

            // If maxEndingHere drops below 0, reset it to 0
            if (maxEndingHere < 0) {
                maxEndingHere = 0;
            }
        }
        return maxSoFar;
    }

    public static void main(String[] args) {
        int[] arr = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
        int maxSum = maxSubArraySum(arr);
        System.out.println("Maximum contiguous subarray sum is: " + maxSum);
    }
}
